proposals/lip_define_state_and_state_transitions_of_dpos_module.md

```diff
@@ -0,0 +1,1472 @@
+    for voter in genesisBlockAssetObject.voters:
+        for sentVote in voter.sentVotes:
+            if sentVote.delegateAddress == validator.address:
+                totalVotesReceived += sentVote.amount
+                if voter.address == validator.address:
+                    selfVotes = sentVote.amount
+  
+    storeKey = validator.address
+    storeValue = {
+        "name": validator.name,
+        "totalVotesReceived": totalVotesReceived,
+        "selfVotes": selfVotes,
+        "lastGeneratedHeight": validator.lastGeneratedHeight,
+        "isBanned": validator.isBanned,
+        "pomHeights": validator.pomHeight,
+        "consecutiveMissedBlocks": validator.consecutiveMissedBlocks
+    } serialized using delegateStoreSchema.
+    ```
+    Further, for every entry `validator` in `genesisBlockAssetObject.validators`, also create an entry in the name substore with 
+    ```python
+    storeKey = validator.name utf8-encoded
+    storeValue = {
+        "delegateAddress": validator.address
+    } serialized using nameStoreSchema.
+    ```
+
+* For each entry `voter` in `genesisBlockAssetObject.voters`, create an entry in the voter substore  with
+    ```python
+    storeKey = uint32be(voter.address)
+    storeValue = {
+        "sentVotes": voter.sentVotes,
+        "pendingUnlocks": voter.pendingUnlocks
+    } serialized using voterStoreSchema.
+    ```
+
+* For each entry `snapshot` in `genesisBlockAssetObject.snapshots`, create an entry in the snapshot substore  with
+    ```python
+    storeKey = uint32be(snapshot.roundNumber)
+    storeValue = {
+        "activeDelegates": snapshot.activeDelegates,
+        "delegateWeightSnapshot": snapshot.delegateWeightSnapshot
+    } serialized using snapshotStoreSchema.
+    ```
+
+* Create an entry in the genesis data substore with
+    ```python
+    storeKey = EMPTY_BYTES
+    storeValue = {
+        "height": block header height of the genesis block,
+        "initRounds": genesisBlockAssetObject.GenesisData.initRounds,
+        "initDelegates": genesisBlockAssetObject.GenesisData.initDelegates
+    } serialized using availableLocalIDStoreSchema.
+    ```
+
+#### Genesis State Finalization
+
+To finalize the state of the genesis block the following logic is executed.  If any step fails, the block is discarded and has no further effect.
+
+As in the previous point, let `genesisBlockAssetBytes` be the `data` bytes included in the block assets for the DPoS module and let `genesisBlockAssetObject` be the deserialization of `genesisBlockAssetBytes` according to the `genesisDPoSStoreSchema` schema, given above. Then:
+
+```python
+# register all validators in the Validators module
+for validator in genesisBlockAssetObject.validators:
+    registerValidatorKeys(
+        validator.address,
+        validator.proofOfPossession,
+        validator.generatorKey,
+        validator.blsKey
+    )
+
+# check that all sentVotes and pendingUnlocks correspond to locked tokens
```

Comment:
As I said in person, another option could be to instead call the lock API in the finalization step. That would avoid this extra check. I guess here it is simple enough that we could do it both ways, so it's more of a consistency thing. Another advantage of calling the API maybe is that if other modules also try to lock tokens you immediately see if the total amount of locked tokens is greater than the amount of tokens (not sure if you do this "global" check in the token module genesis processing)

Reply:
I slightly prefer the way it is written now, as it is less forceful on developers. 

Also, the tokens are locked per module. So here, we compare the votes to tokens locked for the DPoS module.

Reply:
What do you mean with 'less forceful'? We are imposing the same constraint, just expressed differently no?

yes, that was clear to me. But there is another 'global' constraint (global meaning that it involves possibly many modules): the total locked tokens must be lower than maxUint64. Is this checked in the token module independently?

Reply:
yes, the token module checks that there exists less 2**64 of any token.

Reply:
For me the problem really comes from the fact that the Token module is allowed to defined locked tokens, I think we settled that in a past meeting. So I would rather use that feature and only verify here that they are locked.
The other option would be to not allow any locked tokens in the Token module and lock them from the module where they are used.
Overall, I'm not strongly opinionated here, but I prefer the current way of doing it.

Reply:
Thinking about this, for me it looks more "adequate" or "aligned" with our chain architecture to do the locking via API call like Alessandro said. But no strong opinion either.

I just want to confirm that in my understanding the token module can define locked tokens in its genesis (for a given module ID) and at the same time the DPoS module can lock via API without it being an inconsistency, right?

Reply:
I think that in general it is a bit awkward to impose no locked tokens only for specific modules (as you don't know from the token module perspective if these modules are registered or not). So I would say it would either be no locked tokens at all, or any locked tokens (meaning, for any module ID).

Thinking out loud, if we go for the API approach we would then still have the possibility of DPoS tokens being defined "independently" in the token module, unless we first check from DPoS that there are no tokens locked for it (and only then call the API).

Reply:
We concluded to leave the  check as present (tokens locked in Token genesis asset and DPoS module only verifies the locking)